File: Arrays/1218-lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def findLCA(node):
            if not node:
                return (None, 0)
            
            left_node, left_depth = findLCA(node.left)
            right_node, right_depth = findLCA(node.right)

            if left_depth == right_depth:
                return (node, left_depth + 1)
            elif left_depth > right_depth:
                return (left_node, left_depth + 1)
            else:
                return (right_node, right_depth + 1)
        
        return findLCA(root)[0]
     

File: Arrays/1218-lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<TreeNode*, int> findLCA(TreeNode* root) {
        if(root == NULL) return {nullptr, 0};

        auto left = findLCA(root->left);
        auto right = findLCA(root->right);

        if(left.second == right.second) return {root, left.second + 1};
        else if(left.second > right.second) return {left.first, left.second + 1};
        else return {right.first, right.second + 1};
    }

    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        auto ans = findLCA(root);

        return ans.first;
    }
};